src/lib.rs

```diff
@@ -723,8 +551,281 @@ pub unsafe trait Unaligned {
+    ///
+    /// [1] https://doc.rust-lang.org/reference/types/boolean.html
+    unsafe_impl!(bool: AsBytes, Unaligned);
+}
+safety_comment! {
+    /// SAFETY:
+    /// TODO
+    unsafe_impl!(char: AsBytes);
+}
+safety_comment! {
+    /// SAFETY:
+    /// - `AsBytes`: Per the reference [1], `str` has the same layout as `[u8]`,
+    ///   and `[u8]` is `AsBytes`.
+    ///
+    /// [1] https://doc.rust-lang.org/reference/type-layout.html#str-layout
+    unsafe_impl!(str: AsBytes);
+}
+
+safety_comment! {
+    // `NonZeroXxx` is `AsBytes`, but not `FromBytes`.
+    //
+    /// SAFETY:
+    /// `NonZeroXxx` has the same layout as its associated primitive. Since it
+    /// is the same size, this guarantees it has no padding - integers have no
+    /// padding, and there's no room for padding if it can represent all of the
+    /// same values except 0.
+    ///
+    /// TODO(https://github.com/rust-lang/rust/pull/104082): Cite documentation
+    /// that layout is the same as primitive layout.
+    unsafe_impl!(NonZeroU8: AsBytes);
+    unsafe_impl!(NonZeroI8: AsBytes);
+    unsafe_impl!(NonZeroU16: AsBytes);
+    unsafe_impl!(NonZeroI16: AsBytes);
+    unsafe_impl!(NonZeroU32: AsBytes);
+    unsafe_impl!(NonZeroI32: AsBytes);
+    unsafe_impl!(NonZeroU64: AsBytes);
+    unsafe_impl!(NonZeroI64: AsBytes);
+    unsafe_impl!(NonZeroU128: AsBytes);
+    unsafe_impl!(NonZeroI128: AsBytes);
+    unsafe_impl!(NonZeroUsize: AsBytes);
+    unsafe_impl!(NonZeroIsize: AsBytes);
+}
+
+safety_comment! {
+    /// SAFETY:
+    /// The Rust compiler reuses `0` value to represent `None`, so
+    /// `size_of::<Option<NonZeroXxx>>() == size_of::<xxx>()`; see `NonZeroXxx`
+    /// documentation.
+    ///
+    /// TODO(https://github.com/rust-lang/rust/pull/104082): Cite documentation
+    /// for layout guarantees.
+    unsafe_impl!(Option<NonZeroU8>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroI8>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroU16>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroI16>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroU32>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroI32>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroU64>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroI64>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroU128>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroI128>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroUsize>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroIsize>: FromBytes, AsBytes);
+}
+
+safety_comment! {
+    /// SAFETY:
+    /// TODO
+    ///
+    /// TODO(https://github.com/rust-lang/rust/pull/104081): Cite guaranteed
+    /// size and alignment.
+    unsafe_impl!(T: ?Sized => FromBytes for PhantomData<T>);
+    unsafe_impl!(T: ?Sized => AsBytes for PhantomData<T>);
+    unsafe_impl!(T: ?Sized => Unaligned for PhantomData<T>);
+}
+safety_comment! {
+    /// SAFETY:
+    /// `Wrapping<T>` is guaranteed to have the same layout as `T`.
```

Comment:
"... because it is `#[repr(transparent)]` with a single public field"

Reply:
I don't think it's universally agreed that `#[repr(transparent)]` represents a stability guarantee. I'm actually just citing the documentation here: https://doc.rust-lang.org/nightly/core/num/struct.Wrapping.html#layout-1

Reply:
Ah, TIL. Updated.